### ISSUE TITLE
Out of memory operations on huge numpy arrays

### ISSUE THREAD

**Issue by ankurankan** [pgmpy/pgmpy#906]
### Subject of the issue
The size of numpy arrays can get quite huge in `DiscreteFactor` resulting in `MemoryError`. Therefore, we need a way to store these arrays out of memory and do operations on it. 

Ref: #785 

### Possible Solutions:
1. numpy.memmap :- Stores the array in a binary file and allows reading/writing parts of this array. But we will still need to write a wrapper around this for doing efficient operations on these arrays. 
2. scipy.sparse :- Use scipy.sparse to reduce memory usage.
3. blaze

**Comment by yujianll** [comment]
I met the same MemoryError when fitting parameters of a large Bayesian Network. Is there any recommended solution?

**Comment by ankurankan** [comment]
@yujianll Would it be possible for you to share your code? The only possible reason I can think of is that one of the CPDs in your network is too big to fit in memory. Two of the common ways to deal with huge CPDs in Bayesian Networks is parent divorcing or binning of states.

**Comment by yujianll** [comment]
@ankurankan Thanks! The data file is large so it might not be convenient to share with you. But particularly, the model has a CPD for one node that has about 10^9 cells. And probably it is that table that runs out of memory.
For your advice, I get the idea of parent divorcing, but I have one question: assume I already know how should the new node be connected to the existing graph, then how should I define the states of the new node and get the CPD of the new node? I'd really appreciate if you could give me a high level idea or point me to some references.
BTW, I tried to read in a .bif file created by bnlearn package, but it looks like BIFReader cannot read such a large file (about 20 Gb), do you have any recommendations?

**Comment by ankurankan** [comment]
@yujianll So, for the parent divorcing if you are adding an extra variable in the model, you will basically need to compute the values for that variable manually (based on it's parents in the graph) and add that to your dataset. Then you can learn the CPDs using the fit methods.

I can't think of any reason why BIFReader would throw errors (except MemoryError for CPDs that wouldn't fit in memory). Could you share your error traceback?

**Comment by khalibartan** [comment]
@ankurankan  @yujianll I think exception might be coming because the BIFReader loads the whole file in-memory.

**Comment by yujianll** [comment]
@ankurankan Thank you so much! BIFReader didn't give me error traceback, but it took more than 10 hours and still couldn't load the .bif file, and I just terminated it.

**Comment by yujianll** [comment]
@khalibartan Thanks! Yeah that might be the problem if BIFReader tries to read the whole file into memory at once (the .bif file is 25 GB). So is there any way to load such a big file?

**Comment by ankurankan** [comment]
@khalibartan @yujianll If you are not getting a `MemoryError` while reading the file, it means that loading the file isn't really the problem. The problem would then be that the code is probably really slow if it's taking so much time. I will try to see if there's anything that we can optimize and update you.

**Comment by yujianll** [comment]
Thanks!

**Comment by ankurankan** [comment]
@yujianll I have tried to optimize the Reader by parsing CPD values in parallel and have also added an argument to ignore any property data in the file.

But I don't think this is enough for your use case. I have tested it on the full munin network and it takes around 20s to generate the model on my system. Although the munin network is considered massive, it's filesize is only 1.2 MBs. So, I am not sure about the size of your network, or it is a lot of property data? If it's mostly property data, you can try using the `include_properties=False` now while initializing the reader class.

**Comment by yujianll** [comment]
@ankurankan Oh great! Thanks! I haven't try the new reader. I successfully did parameter learning in pgmpy and stored the model through pickle. I'll try the reader.
BTW, I have another question for the Bayesian Model. Why do you use list to store cpds of the model rather than a dictionary? Is dict taking more memory or slower?

**Comment by ankurankan** [comment]
@yujianll Hmm, I am not sure anymore why we went with using a list instead of dict and I can't think of any good reason for that now. I think a dict might be a better option since we access the CPDs by variable quite a bit. I will try to dig deeper and see if dict is a better option. Thanks for pointing this out.

**Comment by Nikhil-Kasukurthi** [comment]
@ankurankan Could you please point to an implementation of the parent divorcing or the approach to implement it? 
Is the algorithm mentioned in this [paper](http://ceur-ws.org/Vol-1419/paper0020.pdf) the one you are referring to?

**Comment by ankurankan** [comment]
@Nikhil-Kasukurthi Yes, that paper proposes a systematic approach. A generally used strategy could be much simpler to look at the variables and see if they can be grouped together.
